taichi/system/traceback.cpp

```diff
@@ -6,25 +6,211 @@
+  RtlCaptureContext(&context);
+
+#if _WIN64
+  STACKFRAME frame = {};
+  frame.AddrPC.Offset = context.Rip;
+  frame.AddrPC.Mode = AddrModeFlat;
+  frame.AddrFrame.Offset = context.Rbp;
+  frame.AddrFrame.Mode = AddrModeFlat;
+  frame.AddrStack.Offset = context.Rsp;
+  frame.AddrStack.Mode = AddrModeFlat;
+#else
+  STACKFRAME frame = {};
+  frame.AddrPC.Offset = context.Eip;
+  frame.AddrPC.Mode = AddrModeFlat;
+  frame.AddrFrame.Offset = context.Ebp;
+  frame.AddrFrame.Mode = AddrModeFlat;
+  frame.AddrStack.Offset = context.Esp;
+  frame.AddrStack.Mode = AddrModeFlat;
+#endif
+
+  bool first = true;
+
+  std::vector<StackFrame> frames;
+  while (StackWalk(machine, process, thread, &frame, &context, NULL,
+                   SymFunctionTableAccess, SymGetModuleBase, NULL)) {
+    StackFrame f = {};
+    f.address = frame.AddrPC.Offset;
+
+#if _WIN64
+    DWORD64 moduleBase = 0;
+#else
+    DWORD moduleBase = 0;
+#endif
+
+    moduleBase = SymGetModuleBase(process, frame.AddrPC.Offset);
+
+    char moduelBuff[MAX_PATH];
+    if (moduleBase &&
+        GetModuleFileNameA((HINSTANCE)moduleBase, moduelBuff, MAX_PATH)) {
+      f.module = basename(moduelBuff);
+    } else {
+      f.module = "Unknown Module";
+    }
+#if _WIN64
+    DWORD64 offset = 0;
+#else
+    DWORD offset = 0;
+#endif
+    char symbolBuffer[sizeof(IMAGEHLP_SYMBOL) + 255];
+    PIMAGEHLP_SYMBOL symbol = (PIMAGEHLP_SYMBOL)symbolBuffer;
+    symbol->SizeOfStruct = (sizeof IMAGEHLP_SYMBOL) + 255;
+    symbol->MaxNameLength = 254;
+
+    if (SymGetSymFromAddr(process, frame.AddrPC.Offset, &offset, symbol)) {
+      f.name = symbol->Name;
+    } else {
+      DWORD error = GetLastError();
+      trace(__FUNCTION__ ": Failed to resolve address 0x%X: %u\n",
+            frame.AddrPC.Offset, error);
+      f.name = "Unknown Function";
+    }
+
+    IMAGEHLP_LINE line;
+    line.SizeOfStruct = sizeof(IMAGEHLP_LINE);
+
+    DWORD offset_ln = 0;
+    if (SymGetLineFromAddr(process, frame.AddrPC.Offset, &offset_ln, &line)) {
+      f.file = line.FileName;
+      f.line = line.LineNumber;
+    } else {
+      DWORD error = GetLastError();
+      trace(__FUNCTION__ ": Failed to resolve line for 0x%X: %u\n",
+            frame.AddrPC.Offset, error);
+      f.line = 0;
+    }
+
+    if (!first) {
+      frames.push_back(f);
+    }
+    first = false;
+  }
+
+  SymCleanup(process);
+
+  return frames;
+}
+
+inline void handle_assert(const char *func, const char *cond) {
+  std::stringstream buff;
+  buff << func << ": Assertion '" << cond << "' failed! \n";
+  buff << "\n";
+
+  std::vector<StackFrame> stack = stack_trace();
+  buff << "Callstack: \n";
+  for (unsigned int i = 0; i < stack.size(); i++) {
+    buff << "0x" << std::hex << stack[i].address << ": " << stack[i].name << "("
+         << std::dec << stack[i].line << ") in " << stack[i].module << "\n";
+  }
+
+  // please replace with std::printf
+  MessageBoxA(NULL, buff.str().c_str(), "Assert Failed", MB_OK | MB_ICONSTOP);
```

Comment:
please replace with `MessageBoxA` with `std::printf`

Reply:
Actually this function is not called... I implemented it in `void print_traceback()`. Shall we delete this function?

Reply:
Oh I see. Please delete this function then. You use Windows more than I do so please feel free to customize it so that it works best for you :-)